Added idempotency key option for charging cards

agrogan opened the pull request:
Hi Jayme,
I have updated a couple files to allow an optional idempotency key to be passed in.  I tested and it's working well to stop a duplicate charge.  This the first time I've done a pull request on github, so I apologize in advance if this isn't the proper way.
Thanks,
Angie

jaymedavis commented:
awesome. :) ... you have to start somewhere!

I will check this out when I get the chance. thanks for the help!

jaymedavis commented:
this will be part of #328 - a new RequestOptions overload for the service methods. You can read up on that if you'd like, it will be a lot cleaner since multiple things can actually be passed to the requestor. Thanks so much for doing the work in this pull request to make it easy for me to see how it works!

agrogan commented:
Great! Thanks for the update, Jayme!

jaymedavis commented:
I must be missing something with this - I have added the header, created a charge and passed an Idempotency-Key, but it is not coming back on the response. Did you have any trouble with this?

agrogan commented:
Hi Jayme,

When I set up my code, I tested with trying to send in a duplicate key, and it failed properly. 

I just checked, and I don't receive it back on the response either, but it seems to be correctly stopping dup keys.  Not sure if this helps, but here are the changes I made. Let me know if you want me to test anything for you.

Thanks,

Angie
_______________________________

I added this function to requestor.cs:

        public static string PostString(string url, string apiKey = null, string idempotencyKey = "")
        {
            var wr = GetWebRequest(url, "POST", apiKey, idempotencyKey);

            return ExecuteWebRequest(wr);
        }

I also updated this function in the Requestor.cs file:

        internal static WebRequest GetWebRequest(string url, string method, string apiKey = null, string idempotencyKey = "", bool useBearer = false)
        {
            apiKey = apiKey ?? StripeConfiguration.GetApiKey();

            var request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = method;

            if(!useBearer)
                request.Headers.Add("Authorization", GetAuthorizationHeaderValue(apiKey));
            else
                request.Headers.Add("Authorization", GetAuthorizationHeaderValueBearer(apiKey));

            request.Headers.Add("Stripe-Version", StripeConfiguration.ApiVersion);
            if (idempotencyKey != "")
            {
                request.Headers.Add("Idempotency-Key", idempotencyKey);
            }
            request.ContentType = "application/x-www-form-urlencoded";
            request.UserAgent = "Stripe.net (https://github.com/jaymedavis/stripe.net)";

            return request;
        }

Then, I updated the StripeChargeService.cs Create function to look like this:

        public virtual StripeCharge Create(StripeChargeCreateOptions createOptions, string idempotencyKey = "")
        {
            var url = this.ApplyAllParameters(createOptions, Urls.Charges, false);

            var response = Requestor.PostString(url, ApiKey, idempotencyKey);

            return Mapper<StripeCharge>.MapFromJson(response);
        }

jaymedavis commented:
I realized it wasn't on the response but stoping dupes as well. This will be in the next release. :)

agrogan commented:
Wonderful! Thanks for all the work you have put into this....it's GREAT!

jaymedavis commented:
You're welcome. :)

I'm still trying to verify how to test the charges are working. I'm pretty sure they are, but they seem to take awhile to show up in the dashboard. I've pinged Stripe to see if there's anything on the response I can use to verify since they aren't in the dashboard yet.

agrogan commented:
Hmm. That's odd.  I just ran a couple unit tests to check mine, and they are showing up right away on the Dashboard.

jaymedavis commented:
Did you create just a charge or customer etc?

agrogan commented:
Just a charge. We are using Stripe for one time charges, so we aren't creating customers.
